rohmu/atomic_opener.py

```diff
@@ -0,0 +1,128 @@
+# about the fact that Write* variables are being redefined.
+# so we need to pour in an additional branch.
+if TYPE_CHECKING:
+    from typing import Literal
+
+    Write = Literal["w"]
+    WriteBinary = Literal["wb"]
+    WriteTest = Literal["somethingrandomw"]
+else:
+    try:
+        from typing import Literal
+
+        Write = Literal["w"]
+        WriteBinary = Literal["wb"]
+        WriteTest = Literal["somethingrandomw"]
+    except ImportError:
+        from typing import Any
+
+        Write = Any
+        WriteBinary = Any
+        WriteTest = Any
+
+
+import errno
+import os
+
+
+def _fd_close_quietly(fd: int) -> None:
+    try:
+        os.close(fd)
+    except OSError as e:
+        if e.errno == errno.EBADF:
+            # closed already
+            return
+        raise
+
+
+@overload
+def atomic_opener(
+    final_path: Path,
+    mode: WriteBinary,
+    encoding: Optional[str] = None,
+    _fd_spy: Callable[[int], None] = lambda unused: None,
+    _after_link_hook: Callable[[], None] = lambda: None,
+) -> ContextManager[BinaryIO]:
+    ...
+
+
+@overload
+def atomic_opener(
+    final_path: Path,
+    mode: Write,
+    encoding: Optional[str] = None,
+    _fd_spy: Callable[[int], None] = lambda unused: None,
+    _after_link_hook: Callable[[], None] = lambda: None,
+) -> ContextManager[TextIO]:
+    ...
+
+
+if TYPE_CHECKING:
+    # necessary for testing or tests will fail type checking
+    @overload
+    def atomic_opener(
+        final_path: Path,
+        mode: WriteTest,
+        encoding: Optional[str] = None,
+        _fd_spy: Callable[[int], None] = lambda unused: None,
+        _after_link_hook: Callable[[], None] = lambda: None,
+    ) -> ContextManager[TextIO]:
+        ...
+
+
+def atomic_opener(
+    final_path: Path,
+    mode: str,
+    encoding: Optional[str] = None,
+    _fd_spy: Callable[[int], None] = lambda unused: None,
+    _after_link_hook: Callable[[], None] = lambda: None,
+) -> ContextManager[Union[TextIO, BinaryIO]]:
+    return _atomic_opener(final_path, mode, encoding, _fd_spy, _after_link_hook)
+
+
+@contextmanager
+def _atomic_opener(
+    final_path: Path,
+    mode: str,
+    encoding: Optional[str] = None,
+    _fd_spy: Callable[[int], None] = lambda unused: None,
+    _after_link_hook: Callable[[], None] = lambda: None,
+) -> Generator[Union[TextIO, BinaryIO], None, None]:
+    """
+    Creates a file object for writing which will only appear on the filesystem if the context manager succeeds.
+
+    It's designed to prevent partial writes (file will either be available with full content, or will be unavailable)
+    and the need for manual cleanup.
+    """
+    parent_dir = final_path.parent
+    if not parent_dir.exists():
+        raise IOError(f"Parent directory '{parent_dir}' must exist but is missing")
+    if "w" not in mode:
+        raise ValueError("Write mode must be used to make actual sense")
+
+    fd = os.open(str(parent_dir), os.O_TMPFILE | os.O_RDWR, 0o600)
```

Comment:
why not use the standard library like it was done with `atomic_create_file`?

It also fixes issues with weird `O_TMPFILE`:

> The above constants are extensions and not present if they are not defined by the C library.

Which would be the case for:

> [..] platforms that are neither Posix nor Cygwin

Reply:
atomic_create_file() is indeed easier, but it pollutes source directory with .tmp files with no chance of automated cleanup, so it requires the caller to be 100% aware of what's happening and perform compensative actions if complete failures (e.g process crash) happen.

> The above constants are extensions and not present if they are not defined by the C library.

It's available since 9 years in the Linux kernel. Does pghoard support anything else than Linux, actually?

Reply:
Thank you for the explanation. Indeed `mkstemp` doesn't delete the file:

> Unlike [TemporaryFile()](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile), the user of [mkstemp()](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp) is responsible for deleting the temporary file when done with it.

I also tested with:

```python
import tempfile, sys, os.path

file_path = "/tmp/test"
fd, tmp_file_path = tempfile.mkstemp(
    prefix=os.path.basename(file_path),
    dir=os.path.dirname(file_path),
    suffix=".metadata_tmp",
)
print(tmp_file_path)
sys.exit(0)  # above file is not deleted!
```

But maybe we can still use `NamedTemporaryFile` from the stdlib? (No strong feelings, now the work has been done anyways, it would just have saved a bit of work with the typing stuff)

```python
import tempfile, sys, os

with tempfile.NamedTemporaryFile() as f:
    print(f.name)
    os.link(f.name, "/tmp/test")
    sys.exit(0)  # temp file is gone, /tmp/test is created
```

> It's available since 9 years in the Linux kernel. Does pghoard support anything else than Linux, actually?

Well, I was just pointing out the advantage listed on the Python docs, and seem easy enough to use. But yeah, I suppose it is not an explicit goal of any of our projects

Reply:
> But maybe we can still use `NamedTemporaryFile` from the stdlib?

sys.exit() is the graceful way to exit Python. It's just like assuming a "finally" block will always be executed. It's got exactly the same issues as mkstemp. What happens if the process exits abruptly? We're left with a file that nobody will ever cleanup - and, during the write operations, which could take time, a temporary file appears in the same directory as the destination - which is something that the client might not expect.

```
>>> with tempfile.NamedTemporaryFile() as f:
...    print(f.name)
...    os._exit(1)
... 
/tmp/tmpu_5wsslx
alan.franzoni@toolbox:~/aiven/aiven-core(dmitry-potepalov-cassandra-fix-rack-distribution)$ ls -l /tmp/tmpu_5wsslx
-rw------- 1 alan.franzoni alan.franzoni 0 Sep 14 16:21 /tmp/tmpu_5wsslx
```

I didn't do this because I wanted; I tried to find other ways.

In order to do this "the right way", we either use O_TMPFILE, or we need to remove ```get_contents_to_file``` from rohmu interface; then a client is forced to properly handle file behaviour as it sees fit. If you people think we should remove the method from the interface and the current implementation has issues, I'll go that way; but just trying to look where rohmu is actually used could take a bit of time, since it's public we should discuss deprecations, etc.

Reply:
> sys.exit() is the graceful way to exit Python.

I always forget it throws an exception 😬

Reply:
@hackaugusto I tried with TemporaryFile:
```
    with TemporaryFile(dir=parent_dir, mode=mode, encoding=encoding) as file_obj:
        if "b" in mode:
            yield cast(BinaryIO, file_obj)
        else:
            yield cast(TextIO, file_obj)
        file_obj.flush()
        path = "/proc/self/fd/{0}".format(file_obj.fileno())
        assert Path(path).exists()
        os.link(path, str(final_path), src_dir_fd=0, follow_symlinks=True)
        _after_link_hook()
```

But this seems to fail:
```
            path = "/proc/self/fd/{0}".format(file_obj.fileno())
            assert Path(path).exists()
>           os.link(path, str(final_path), src_dir_fd=0, follow_symlinks=True)
E           FileNotFoundError: [Errno 2] No such file or directory: '/proc/self/fd/11' -> '/tmp/pytest-of-alan.franzoni/pytest-17/test_file_is_atomically_create0/something'
```

So, it seems we get a different set of problems by using python's own TemporaryFile, but I think this "atomic open" function would be a nice addition to python's own stdlib and I'll try contributing it upstream, if it's possible to make this work.

Reply:
So, I tried investigating since I was at it; it seems that Python uses O_EXCL when opening a TemporaryFile, and this prevents (implementation choice) to actually link in the filesystem. I think this is more incidental than intentional, though.

Reply:
@alanfranz thanks for looking into it :bow: